sync2/multipeer/multipeer.go

```diff
@@ -0,0 +1,413 @@
+				if errors.Is(err, context.Canceled) {
+					return err
+				}
+			} else {
+				probeCh <- probeResult{p, pr}
+			}
+			return nil
+		})
+	}
+
+	var egConsume errgroup.Group
+	egConsume.Go(func() error {
+		for pr := range probeCh {
+			c, err := mpr.syncBase.Count()
+			if err != nil {
+				return err
+			}
+
+			// We do not consider peers with substantially fewer items than the local
+			// set for active sync. It's these peers' responsibility to request sync
+			// against this node.
+			if pr.Count+mpr.cfg.MaxSyncDiff < c {
+				mpr.logger.Debug("skipping peer with low item count",
+					zap.Int("peerCount", pr.Count),
+					zap.Int("localCount", c))
+				continue
+			}
+
+			s.syncable = append(s.syncable, pr.p)
+			if pr.Count > mpr.cfg.MinSplitSyncCount {
+				mpr.logger.Debug("splitSyncable peer",
+					zap.Stringer("peer", pr.p),
+					zap.Int("count", pr.Count))
+				s.splitSyncable = append(s.splitSyncable, pr.p)
+			} else {
+				mpr.logger.Debug("NOT splitSyncable peer",
+					zap.Stringer("peer", pr.p),
+					zap.Int("count", pr.Count))
+			}
+
+			mDiff := float64(mpr.cfg.MaxFullDiff)
+			if math.Abs(float64(pr.Count-c)) < mDiff && (1-pr.Sim)*float64(c) < mDiff {
+				mpr.logger.Debug("nearFull peer",
+					zap.Stringer("peer", pr.p),
+					zap.Float64("sim", pr.Sim),
+					zap.Int("localCount", c))
+				s.nearFullCount++
+			} else {
+				mpr.logger.Debug("nearFull peer",
+					zap.Stringer("peer", pr.p),
+					zap.Float64("sim", pr.Sim),
+					zap.Int("localCount", c))
+			}
+		}
+		return nil
+	})
+	err := eg.Wait()
+	close(probeCh)
+	if err != nil {
+		egConsume.Wait()
+		return s, err
+	}
+	return s, egConsume.Wait()
+}
+
+func (mpr *MultiPeerReconciler) needSplitSync(s syncability) bool {
+	if float64(s.nearFullCount) >= float64(len(s.syncable))*mpr.cfg.MinCompleteFraction {
+		// enough peers are close to this one according to minhash score, can do
+		// full sync
+		mpr.logger.Debug("enough peers are close to this one, doing full sync",
+			zap.Int("nearFullCount", s.nearFullCount),
+			zap.Int("peerCount", len(s.syncable)),
+			zap.Float64("minCompleteFraction", mpr.cfg.MinCompleteFraction))
+		return false
+	}
+
+	if len(s.splitSyncable) < mpr.cfg.MinSplitSyncPeers {
+		// would be nice to do split sync, but not enough peers for that
+		mpr.logger.Debug("not enough peers for split sync",
+			zap.Int("splitSyncableCount", len(s.splitSyncable)),
+			zap.Int("minSplitSyncPeers", mpr.cfg.MinSplitSyncPeers))
+		return false
+	}
+
+	mpr.logger.Debug("can do split sync")
+	return true
+}
+
+func (mpr *MultiPeerReconciler) fullSync(ctx context.Context, syncPeers []p2p.Peer) error {
+	var eg errgroup.Group
+	for _, p := range syncPeers {
+		syncer := mpr.syncBase.Derive(p)
+		eg.Go(func() error {
+			err := syncer.Sync(ctx, nil, nil)
+			switch {
+			case err == nil:
+				mpr.sl.NoteSync()
+			case errors.Is(err, context.Canceled):
+				syncer.Release()
```

Comment:
That's how it used to be, but then there was a comment above about `Release()` being called twice (even though that's safe), so I updated it to avoid this situation :)
(we need the error code from `Release()` below)